File: web/backend/app/mbtiles_server.py
from __future__ import annotations

import sqlite3
from pathlib import Path

from terrain_converter.mbtiles import xyz_to_tms_row

from .models import BBox, MapView


MEDIA_TYPES = {
    "png": "image/png",
    "jpg": "image/jpeg",
    "jpeg": "image/jpeg",
    "webp": "image/webp",
    "pbf": "application/x-protobuf",
}


def _sniff_media_type(tile_data: bytes) -> str:
    if tile_data.startswith(b"\x89PNG"):
        return "image/png"
    if tile_data.startswith(b"\xff\xd8\xff"):
        return "image/jpeg"
    if tile_data.startswith(b"RIFF") and tile_data[8:12] == b"WEBP":
        return "image/webp"
    return "application/octet-stream"


class MBTilesServer:
    def __init__(self, path: Path) -> None:
        self.path = path

    def exists(self) -> bool:
        return self.path.exists()

    def get_metadata(self) -> dict[str, str]:
        if not self.path.exists():
            return {}
        connection = sqlite3.connect(self.path)
        try:
            rows = connection.execute("SELECT name, value FROM metadata").fetchall()
        finally:
            connection.close()
        return {name: value for name, value in rows}

    def get_format(self) -> str:
        return self.get_metadata().get("format", "png").lower()

    def get_name(self) -> str | None:
        return self.get_metadata().get("name")

    def get_attribution(self) -> str | None:
        return self.get_metadata().get("attribution")

    def get_minzoom(self) -> int | None:
        value = self.get_metadata().get("minzoom")
        if value is None:
            return None
        try:
            return int(float(value))
        except ValueError:
            return None

    def get_maxzoom(self) -> int | None:
        value = self.get_metadata().get("maxzoom")
        if value is None:
            return None
        try:
            return int(float(value))
        except ValueError:
            return None

    def get_bounds(self) -> BBox | None:
        value = self.get_metadata().get("bounds")
        if not value:
            return None
        try:
            west, south, east, north = [float(item) for item in value.split(",", 3)]
        except ValueError:
            return None
        return BBox(west=west, south=south, east=east, north=north)

    def get_view(self) -> MapView | None:
        metadata = self.get_metadata()
        center = metadata.get("center")
        if center:
            try:
                center_lon, center_lat, zoom = center.split(",", 2)
                return MapView(center_lon=float(center_lon), center_lat=float(center_lat), zoom=int(float(zoom)))
            except ValueError:
                pass

        bounds = self.get_bounds()
        if bounds is None:
            return None
        minzoom = metadata.get("minzoom", "0")
        try:
            zoom = int(float(minzoom))
        except ValueError:
            zoom = 0
        return MapView(
            center_lon=(bounds.west + bounds.east) / 2,
            center_lat=(bounds.south + bounds.north) / 2,
            zoom=zoom,
        )

    def get_tile(self, z: int, x: int, y: int) -> tuple[bytes, str] | None:
        if not self.path.exists():
            return None
        connection = sqlite3.connect(self.path)
        try:
            fmt_row = connection.execute("SELECT value FROM metadata WHERE name='format'").fetchone()
            row = connection.execute(
                "SELECT tile_data FROM tiles WHERE zoom_level=? AND tile_column=? AND tile_row=?",
                (z, x, xyz_to_tms_row(z, y)),
            ).fetchone()
        finally:
            connection.close()
        if row is None:
            return None
        tile_data = row[0]
        media_type = MEDIA_TYPES.get((fmt_row[0] if fmt_row else "").lower(), _sniff_media_type(tile_data))
        return tile_data, media_type

File: kotlin/terrain-web/src/main/kotlin/com/terrainconverter/web/MbtilesServer.kt
package com.terrainconverter.web

import com.terrainconverter.core.xyzToTmsRow
import java.nio.file.Path
import java.sql.Connection
import java.sql.DriverManager
import kotlin.io.path.exists

private val mediaTypes = mapOf(
    "png" to "image/png",
    "jpg" to "image/jpeg",
    "jpeg" to "image/jpeg",
    "webp" to "image/webp",
    "pbf" to "application/x-protobuf",
)

private fun sniffMediaType(tileData: ByteArray): String {
    return when {
        tileData.copyOfRange(0, minOf(tileData.size, 4)).contentEquals(byteArrayOf(0x89.toByte(), 0x50, 0x4E, 0x47)) -> "image/png"
        tileData.size >= 3 && tileData[0] == 0xFF.toByte() && tileData[1] == 0xD8.toByte() && tileData[2] == 0xFF.toByte() -> "image/jpeg"
        tileData.size >= 12 && String(tileData.copyOfRange(0, 4)) == "RIFF" && String(tileData.copyOfRange(8, 12)) == "WEBP" -> "image/webp"
        else -> "application/octet-stream"
    }
}

class MBTilesServer(private val path: Path) {
    init {
        Class.forName("org.sqlite.JDBC")
    }

    fun exists(): Boolean = path.exists()

    private fun <T> withConnection(block: (Connection) -> T): T {
        return DriverManager.getConnection("jdbc:sqlite:${path.toAbsolutePath()}").use(block)
    }

    fun getMetadata(): Map<String, String> {
        if (!exists()) return emptyMap()
        return withConnection { connection ->
            connection.prepareStatement("SELECT name, value FROM metadata").use { statement ->
                statement.executeQuery().use { rs ->
                    linkedMapOf<String, String>().apply {
                        while (rs.next()) {
                            put(rs.getString(1), rs.getString(2))
                        }
                    }
                }
            }
        }
    }

    fun getFormat(): String = getMetadata()["format"]?.lowercase() ?: "png"

    fun getName(): String? = getMetadata()["name"]

    fun getAttribution(): String? = getMetadata()["attribution"]

    fun getMinzoom(): Int? = getMetadata()["minzoom"]?.toDoubleOrNull()?.toInt()

    fun getMaxzoom(): Int? = getMetadata()["maxzoom"]?.toDoubleOrNull()?.toInt()

    fun getBounds(): BBox? {
        val parts = getMetadata()["bounds"]?.split(",") ?: return null
        if (parts.size != 4) return null
        val values = parts.map { it.toDoubleOrNull() ?: return null }
        return BBox(values[0], values[1], values[2], values[3])
    }

    fun getView(): MapView? {
        val metadata = getMetadata()
        metadata["center"]?.split(",")?.let { parts ->
            if (parts.size == 3) {
                val lon = parts[0].toDoubleOrNull()
                val lat = parts[1].toDoubleOrNull()
                val zoom = parts[2].toDoubleOrNull()?.toInt()
                if (lon != null && lat != null && zoom != null) {
                    return MapView(lon, lat, zoom)
                }
            }
        }
        val bounds = getBounds() ?: return null
        return MapView(
            centerLon = (bounds.west + bounds.east) / 2.0,
            centerLat = (bounds.south + bounds.north) / 2.0,
            zoom = metadata["minzoom"]?.toDoubleOrNull()?.toInt() ?: 0,
        )
    }

    fun getTile(z: Int, x: Int, y: Int): Pair<ByteArray, String>? {
        if (!exists()) return null
        return withConnection { connection ->
            val fmt = connection.prepareStatement("SELECT value FROM metadata WHERE name='format'").use { statement ->
                statement.executeQuery().use { rs -> if (rs.next()) rs.getString(1) else null }
            }
            connection.prepareStatement(
                "SELECT tile_data FROM tiles WHERE zoom_level=? AND tile_column=? AND tile_row=?"
            ).use { statement ->
                statement.setInt(1, z)
                statement.setInt(2, x)
                statement.setInt(3, xyzToTmsRow(z, y))
                statement.executeQuery().use { rs ->
                    if (!rs.next()) return@withConnection null
                    val tileData = rs.getBytes(1)
                    tileData to (mediaTypes[fmt?.lowercase()] ?: sniffMediaType(tileData))
                }
            }
        }
    }
}
